### ISSUE TITLE
Throttle on only a single route

### ISSUE THREAD

**Issue by milad-afkhami** [nestjs/throttler#1900]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe it

My challenge is how to apply throttling to a specific route without setting up a global throttle. My concern arises from the necessity to log every single request from every single user (in memory) when establishing a global throttling system, to ascertain whether the new request necessitates throttling. How can I accomplish this?

### Describe the solution you'd like

Globally config it like:
```js
imports: [
    ...
    ThrottlerModule.forRoot()
]
```

and use it like this on the route:
```js
@Throttle([{ limit: 3, ttl: 60000 }])
@Get()
findAll() {
  return "List users works with custom rate limiting.";
}
```

Notes:

1. It's better to remove the default keyword on the Throttle decorator
2. It's good to add support for arrays inside the Throttle decorator
3. I have tested the global config and added a guard with an empty config and it doesn't work.

### What is the motivation / use case for changing the behavior?

For more customizations and more control for throttling and saving more space.

**Comment by milad-afkhami** [comment]
@canoir it was your company's concern too, right?

**Comment by jmcdo29** [comment]
Okay, I think I see the request here. Instead of being forced to use the `ThrottlerModule` it would be nice to set the `@UseGuards()` and `@Throttle()` on a case-by-case basis, correct? 

I think the `ThrottlerModule` should still be imported for the access to the storage service, but that may be movable to not be required in a dynamic module, or we just keep the `forRoot` as is and allow an empty function call (easiest as there's no changes there)

In the `@Throttle()`, I can probably allow it to take in the config or a record of `name: config` for sake of ease and transform it under the hood. If we were to take in either the record or the array, if the array had two configurations, both without name, the latter would overwrite the former and lead to unexpected behavior. Do you think this would make for a good DX?

---

As for the initial "How can I implement this?", so long as there's a config object passed to the `forRoot()` (something that has a `ttl` and `limit`, it can be ridiculous numbers too, e.g. `[{ limit: 0, ttl: 0 }]`) the n the guard should work with the overwrite from the `@Throttle()` decorator

**Comment by Canoir** [comment]
> Okay, I think I see the request here. Instead of being forced to use the `ThrottlerModule` it would be nice to set the `@UseGuards()` and `@Throttle()` on a case-by-case basis, correct?
> 
> I think the `ThrottlerModule` should still be imported for the access to the storage service, but that may be movable to not be required in a dynamic module, or we just keep the `forRoot` as is and allow an empty function call (easiest as there's no changes there)
> 
> In the `@Throttle()`, I can probably allow it to take in the config or a record of `name: config` for sake of ease and transform it under the hood. If we were to take in either the record or the array, if the array had two configurations, both without name, the latter would overwrite the former and lead to unexpected behavior. Do you think this would make for a good DX?
> 
> As for the initial "How can I implement this?", so long as there's a config object passed to the `forRoot()` (something that has a `ttl` and `limit`, it can be ridiculous numbers too, e.g. `[{ limit: 0, ttl: 0 }]`) the n the guard should work with the overwrite from the `@Throttle()` decorator

I agree with you and about the name: config , I think that is great too, our need is exactly what you said, but your other answer make me a bigger question, what will happen if we pass ttl: 0 and limit 0 to global config?  I mean if the guard only passes cause of 0 then there is a little bit shitty but solution to our problem even now, right?

**Comment by jmcdo29** [comment]
The idea I suggested was just to make sure that there _was_ a default config so that the `@Throttle()` override properly works. So long as the guard is not applied to any routes _without_ the `@Throttle()` then the point is moot,. If it is, you'll find out pretty quickly :smile:

**Comment by flazouh** [comment]
Hi guys is this the right types declaration ?

```ts
export declare const Throttle: (limit?: number, ttl?: number) => MethodDecorator & ClassDecorator;
export declare const SkipThrottle: (skip?: boolean) => MethodDecorator & ClassDecorator;
export declare const InjectThrottlerOptions: () => PropertyDecorator & ParameterDecorator;
export declare const InjectThrottlerStorage: () => PropertyDecorator & ParameterDecorator;

```

I see everywhere in the doc @Throttle({limit: 10, ttl: 60000}) or @Throttle(default: {limit: 10, ttl: 60000}) but the types I have is actually a function that take 2 arguments 🤔

**Comment by mukunda-** [comment]
It took me an hour to realize that I *need* to have a default config block, even if I'm not using it. I'd say it should have a default if not specified.

**Comment by dominic-schmid** [comment]
So, I'm not sure if I'm just doing something wrong but:

- [x]  Global throttling configuration with a global  `ThrottlerGuard` provider works
- [ ] Global throttling configuration + overriding controller method (ignored)
- [ ] Global throttling configuration + overriding controller method + providing `ThrottlerGuard` specifically for module (ignored)

How would I mark only certain routes to be throttled then?
`@Throttle({ long: { limit: 5, ttl: 60000 } })`

**Comment by mukunda-** [comment]
@dominic-schmid currently it would be like this:
1. import throttler with a default configuration provided
2. mark your route with the `@Throttle` line and a `@UseGuards(ThrottlerGuard)`

For all routes to be affected, there is another way to apply the ThrottlerGuard globally. Otherwise, it is only going to affect specific routes marked with the ThrottlerGuard.

This issue is concerning step 1 - it would be better if there was a preconfigured "default" rather than needing to specify a dummy configuration if you are using custom limit/ttl for each route.

**Comment by jvvcn** [comment]
Having same problem - I do use module with default configuration, however there are routes that I would like to lower the rate limit. Currently I cannot do it without overriding default configurations... Any ideas how to workaround it?

**Comment by jmcdo29** [comment]
> Having same problem - I do use module with default configuration, however there are routes that I would like to use different rate limit that is defined in module. Currently I cannot do it without overriding default configurations... Any ideas how to workaround it?

@jvvcn that's kind of the idea of what should be done. Is there something you'd rather be able to do?

**Comment by jameslufz** [comment]
I have the same problem and need like you do.

I haven't found any solutions for how to use Throttle without global setup, but I have an alternative for you.

In `app.module.ts` you can set conditions for using Throttle for some routes by `skipIf` method.

``` typescript
import { ThrottlerGuard, ThrottlerModule } from '@nestjs/throttler';
import { Request } from 'express';

@Module({
    imports: [
        ThrottlerModule.forRoot([{
            ttl: (1000 * 60 * 5),
            limit: 5,
            skipIf: (ctx: ExecutionContext) =>
            {
                // Here you can.
                const request = ctx.switchToHttp().getRequest() as Request
                return request.path !== '/api/v1/auth/check-something'

                /**
                 * --------------------------------------------
                 * Alternative, use regular expression to make condition for multiple routes.
                 * --------------------------------------------
                 * 
                 * return !/check-cat|check-duck|check-something/gi.test(request.path)
                 */
            }
        }]),
    ],
    controllers: [
        AppController,
    ],
    providers: [
        {
            provide: APP_GUARD,
            useClass: ThrottlerGuard
        },
    ],
})

export class AppModule {}
```

If your target route hasn't too many, I think this way may help you.

**Comment by andrejsoucek** [comment]
@jameslufz This will work however every response will contain `x-ratelimit-*` headers with static numbers, also with `@Throttle()` decorator we lose the possibility to inject configuration.

@jmcdo29 We need an opposite behavior - not enabling the throttler globally but only marking routes which should be throttled. Registering globally is fine, but we need some logic to toggle between current opt-out solution (SkipThrottle decorator) and a new opt-in solution (only throttle decorated routes)

**Comment by lukepolo** [comment]
it would be nice todo this per module configuration , i added a base config and did not import the throttler except for my single route 

```
  @UseGuards(ThrottlerGuard)
  @Throttle({ default: { limit: 1, ttl: 59 * 60 * 1000 } })
  @Post()
  public async telemetry(@Req() request: Request) {
    await this.telemetryService.record(request.body);
  }
```

then in my app

```
ThrottlerModule.forRoot({
      throttlers: [
        {
          ttl: 0,
          limit: 0,
        },
      ],
    }),
```

Notice I did not use
```
{
    provide: APP_GUARD,
    useClass: ThrottlerGuard
},
```  


which headers reply

![Image](https://github.com/user-attachments/assets/3e51a520-f49a-4436-87a4-3d2a6c8c8152)

and the endpoint rate limited

![Image](https://github.com/user-attachments/assets/4e20448c-a16b-48fd-85ca-e039c189a3da)

**Comment by Sharad-Zooper** [comment]
Why this decorator is not taking the ttl and limit which I define in config ?. Why do I need to define them again in decorator ?@Throttle({ strict: {} })
or there any other way to use name throttler as defined in config during forRoot()  or .forRootAsync as below ?
ThrottlerModule.forRootAsync({
  imports: [ConfigModule],
  inject: [ConfigService],
  useFactory: (config: ConfigService) => [
    {
      name: 'strict',
      ttl: config.get<number>('STRICT_TTL', 300000),    // 5 minutes
      limit: config.get<number>('STRICT_LIMIT', 3),     // 3 requests
    },
    // ... other configurations
  ],
}

**Comment by MauriceArikoglu** [comment]
> Why this decorator is not taking the ttl and limit which I define in config ?. Why do I need to define them again in decorator ?[@Throttle](https://github.com/Throttle)({ strict: {} }) or there any other way to use name throttler as defined in config during forRoot() or .forRootAsync as below ? ThrottlerModule.forRootAsync({ imports: [ConfigModule], inject: [ConfigService], useFactory: (config: ConfigService) => [ { name: 'strict', ttl: config.get('STRICT_TTL', 300000), // 5 minutes limit: config.get('STRICT_LIMIT', 3), // 3 requests }, // ... other configurations ], }

wondering about the same thing

**Comment by jmcdo29** [comment]
@copamurp I am the designer and maintainer of this library, would you like to try to build it yourself instead?

There should be no need to overwrite the default configuration from the `forRoot` **unless** you need to _change_ the configuration for that endpoint. Perhaps I can help clarify this in the documentation, but the idea is that you can set the default configuration(s) in the `forRoot` and then use the `ThrottlerGuard` either globally (if you want to rate limit each endpoint) or **only** on the endpoints you want limited.

**Comment by copamurp** [comment]
> @copamurp I am the designer and maintainer of this library, would you like to try to build it yourself instead?
> 
> There should be no need to overwrite the default configuration from the `forRoot` **unless** you need to _change_ the configuration for that endpoint. Perhaps I can help clarify this in the documentation, but the idea is that you can set the default configuration(s) in the `forRoot` and then use the `ThrottlerGuard` either globally (if you want to rate limit each endpoint) or **only** on the endpoints you want limited. 

No, I wouldn't like to design my own rate limiting framework, that's the entire point of being able to use one that's being offered publicly. If I have to design one that's more intuitive than yours, I will. Thanks for your help.

**Comment by jmcdo29** [comment]
Then perhaps rather than saying it is just "unintuitive" you provide feedback on how to make it clearer and easier for others to use. Sometimes as maintainers when we develop things we understand them on a deeper level and take ideas about the libraries for granted that others don't quite grasp yet. I'd be more than happy to try and clarify anything I can or provide a better developer experience, but I do not appreciate you saying that I "need an evaluation"

**Comment by MauriceArikoglu** [comment]
Open Source is about collaboration.

Demands can be made when paying for a service.

So maybe check your entitlement and suggest functionality by implementing it and opening a pull request instead of complaining to someone who invests his / her spare time to make this free software for you to use.

Or maybe just pay - then you can make demands. Lol.

**Comment by MauriceArikoglu** [comment]
> > Open Source is about collaboration.
> > 
> > Demands can be made when paying for a service.
> > 
> > So maybe check your entitlement and suggest functionality by implementing it and opening a pull request instead of complaining to someone who invests his / her spare time to make this free software for you to use.
> > 
> > Or maybe just pay - then you can make demands. Lol.
> 
> I'm entitled to criticize whatever product I want, free or paid. If you're offering a free solution to a problem for developers to use, then you can expect to receive feedback when that solution does not easily apply to the problem it's trying to solve.

And we are free to ignore you. Good day to you.

On a site note, thanks to @jmcdo29 for implementing this library. Our workaround I detailed above works splendid for a >300 endpoint API.

**Comment by copamurp** [comment]
> > > Open Source is about collaboration.
> > > 
> > > Demands can be made when paying for a service.
> > > 
> > > So maybe check your entitlement and suggest functionality by implementing it and opening a pull request instead of complaining to someone who invests his / her spare time to make this free software for you to use.
> > > 
> > > Or maybe just pay - then you can make demands. Lol.
> > 
> > I'm entitled to criticize whatever product I want, free or paid. If you're offering a free solution to a problem for developers to use, then you can expect to receive feedback when that solution does not easily apply to the problem it's trying to solve.
> 
> And we are free to ignore you. Good day to you.
> 
> On a site note, thanks to @jmcdo29 for implementing this library. Our workaround I detailed above works splendid for a >300 endpoint API.

What an excellent response from a member of the open source community. Good day to you and this framework owner as well, I'll be implementing my own solution. It wasn't until now I saw that this original issue was from 2 years ago, which seals the deal for me.

**Comment by MarcOlivierRodrigue** [comment]
In my tests, (`"@nestjs/throttler": "=6.5.0"`) if I set both `ttl` and `limit` to 0 in the global config, the `x-ratelimit-remaining` does not respect the `ttl` for the endpoints where I overrode the global config. As in, it doesn't clear the requests after the `ttl` expires. It only works if I give the global config a dummy `limit` and keep the global `ttl` to 0...

**Comment by mhalikosen** [comment]
I went through the thread and tested each scenario against 6.7.0 (Nest 12.0.3, Node 24.21).

The capability asked for here exists today. No `APP_GUARD` is needed: the module is global, and `@UseGuards(ThrottlerGuard)` on a route limits that route only.

| setup | result |
| --- | --- |
| `forRoot([{ ttl: 60000, limit: 2 }])` with a global guard | `200,200,429,429` |
| no global guard, guard on one route | only that route is limited |
| `@Throttle({ default: {} })` | inherits the `forRoot` values |
| `@Throttle({ default: { limit: 5 } })` | limit 5, `ttl` inherited |
| `@SkipThrottle({ default: true })` | unlimited, no rate limit headers |
| `strict` plus `loose` configured, guard on one route | that route is limited by both |
| `forRoot([])` plus a route override | every request passes, nothing is logged |

The definitions in `forRoot` are one policy rather than a menu, so every entry applies wherever the guard is bound. A route narrows that by skipping a name, which is one decorator line. The override inheriting from `forRoot` is the intended behaviour, and it does inherit. The one real exception is `limit: 0`: `||` in the guard replaces it with the configured value (#2693, PR #2694).

What is missing is documentation, not an API. A README section for the route-level setup is in #2700, and a boot warning for the empty configuration follows as a second PR. Once those are in I think this can close.

**Comment by mhalikosen** [comment]
@jmcdo29 this can close: #2700 and #2701 are merged.

- The README documents throttling specific routes: the guard bound per route, `@Throttle()` inheriting from `forRoot`, and skipping a throttler by name.
- The guard warns at startup when no throttler is configured. It ships with the next release (#2706), and #2707 adds the warning to the README.
- The `limit: 0` exception from my earlier comment is fixed by #2694.

**Comment by kamilmysliwiec** [comment]
This is now documented: see [Throttling specific routes](https://github.com/nestjs/throttler#throttling-specific-routes) in the README (#2700, #2707). In short: skip the `APP_GUARD` provider and bind `@UseGuards(ThrottlerGuard)` only on the routes or controllers that should be limited. `@Throttle({ default: { ... } })` changes only the fields you list and inherits the rest from `forRoot()`.

`ThrottlerModule.forRoot()`/`forRoot([])` with no throttlers also logs a startup warning now (#2701), since that setup silently limits nothing. Closing, thanks everyone for the discussion.
